topics/S3/etcd-migration/src/steps.py

```diff
@@ -0,0 +1,196 @@
+    # add a wait for drain
+    print(f"{bcolors.CYAN}\tWaiting for host {global_vars.s3_hosts[0]} to be drained. if this takes too long, cancel "
+          f"script (^C) fix what is needed and re-run script{bcolors.ENDC}")
+    wait_for_drain(global_vars.s3_hosts[0])
+    print(f"{bcolors.GREEN}\tHost {global_vars.s3_hosts[0]} is in Drain Mode.{bcolors.ENDC}")
+
+
+@step("Migrate all of ETCD data into KWAS")
+def migrate_data():
+    migration_status = get_migration_status(global_vars.s3_hosts[0])
+    if migration_status["in_progress"]:
+        print(f"{bcolors.RED}ERROR: migration is in progress. Fix what is needed and rerun script{bcolors.ENDC}")
+        exit(1)
+    data = {"kwas_endpoint": global_vars.socket_path}
+    r = send_request(f"http://{global_vars.s3_hosts[0]}:9001/minio/migration/migrate/kwas", data)
+    while True:
+        migration_status = get_migration_status(global_vars.s3_hosts[0])
+        if not migration_status["in_progress"]:
+            break
+    if migration_status["error"] != "":
+        print(f"{bcolors.RED}ERROR: migration failed during step {migration_status['step']} with error: "
+              f"'{migration_status['error']}'. Fix what is needed and rerun script{bcolors.ENDC}")
+        exit(1)
+    print(f"{bcolors.GREEN}\tFinished migrating data.{bcolors.ENDC}")
+
+
+@step("Validating migration")
+def validate_migration():
+    data = {"kwas_endpoint": global_vars.socket_path}
+    r = send_request(f"http://{global_vars.s3_hosts[0]}:9001/minio/migration/validate/kwas", data).json()
+    if not r['verified']:
+        print(f"{bcolors.RED}ERROR: Migration Validation failed. Fix it and run script again. The invalid data is: {bcolors.ENDC}")
+        print_errors_dict(r["errors"])
+        exit(1)
+    print(f"{bcolors.GREEN}\tAll migrated data was verified successfully{bcolors.ENDC}")
+
+
+@step("Drain all S3 hosts and Restart them into new mode")
+def drain_and_restarts_all_hosts():
+    for host in global_vars.s3_hosts:
+        if global_vars.auto_mode < Automode.AUTO.value:
+            if not continue_to_next_host():
+                do_revert()
+
+        set_drain_mode(True, int(''.join(filter(str.isdigit, host))))
+        print(f"{bcolors.CYAN}\tWaiting for host {host} to be drained. if this takes too long, cancel "
+              f"script (^C) fix what is needed and re-run script{bcolors.ENDC}")
+        wait_for_drain(host)
+        print(f"{bcolors.DARK_GREEN}\tHost {host} is in Drain Mode, Restarting it into new mode{bcolors.ENDC}")
+
+        output = send_bash_command(f"weka debug manhole --slot 0 --host {int(''.join(filter(str.isdigit, host)))} "
+                                   f"s3_update_config force_minio_refresh=true")
+        while True:
+            time.sleep(2)
+            output = send_bash_command("weka s3 cluster status -J")
+            hosts_health = json.loads(output)
+            if all(val for val in hosts_health.values()):
+                break
+
```

Comment:
i think that instead of doing "restart all minios" step and then "check all minios" step, it's better to also check each minio after its restart (here).

so here i would add a check for that minio is still in read-only migration mode, and that the kwas client (or etcd client during revert) is used, (and optionally that that minio log looks fine tho i understand why it's tricky).